core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/RowDenormalizerAggregator.java

```diff
@@ -0,0 +1,133 @@
+  private String keyField;
+
+  public RowDenormalizerAggregator(RowDenormalizerConfig conf) {
+    this.conf = conf;
+  }
+
+  @Override
+  public void prepareRun(BatchAggregatorContext context) throws Exception {
+    if (conf.numPartitions != null) {
+      context.setNumPartitions(conf.numPartitions);
+    }
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    validateInputFields(pipelineConfigurer.getStageConfigurer().getInputSchema());
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(initializeOutputSchema());
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    outputFields = conf.getOutputFields();
+    outputMappings = conf.getOutputFieldMappings();
+    keyField = conf.getKeyField();
+    outputSchema = initializeOutputSchema();
+  }
+
+  @Override
+  public void groupBy(StructuredRecord record, Emitter<String> emitter) throws Exception {
+    emitter.emit(record.get(keyField).toString());
+  }
+
+  @Override
+  public void aggregate(String groupKey, Iterator<StructuredRecord> iterator,
+                        Emitter<StructuredRecord> emitter) throws Exception {
+    if (!iterator.hasNext()) {
+      return;
+    }
+    StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
+    builder.set(conf.getKeyField(), groupKey);
+    while (iterator.hasNext()) {
+      StructuredRecord record = iterator.next();
+      String fieldName = record.get(conf.getFieldName());
+      fieldName = outputMappings.containsKey(fieldName) ? outputMappings.get(fieldName) : fieldName;
+      if (outputFields.contains(fieldName)) {
+        builder.set(fieldName, record.get(conf.getFieldValue()));
+      }
+    }
+    emitter.emit(builder.build());
+  }
+
+  /**
+   * Builds the output schema, using output fields provided by user.
+   */
+  private Schema initializeOutputSchema() {
+    List<Schema.Field> fields = new ArrayList<>();
+    // Input fields coming from source will be of type String; that's why we are building the output schema as String.
+    fields.add(Schema.Field.of(conf.getKeyField(), Schema.of(Schema.Type.STRING)));
+    for (String outputField : conf.getOutputFields()) {
+      fields.add(Schema.Field.of(outputField, Schema.nullableOf(Schema.of(Schema.Type.STRING))));
+    }
+    return Schema.recordOf("output.schema", fields);
+  }
+
+  /**
+   * @param inputSchema Validates whether the keyfield, fieldname, and fieldvalue entered by the user is present in
+   *                    the input schema or not.
+   * @return true
+   */
+  private void validateInputFields(Schema inputSchema) {
+    if ((inputSchema.getField(conf.getKeyField())) == null) {
+      throw new IllegalArgumentException(
+        String.format("Keyfield '%s' does not exist in input schema %s", conf.getKeyField(), inputSchema));
+    } else if ((inputSchema.getField(conf.getFieldName())) == null) {
+      throw new IllegalArgumentException(
```

Comment:
need to check that the field is of type string. Similar thing, is this allowed to be null? If so, then a null means that it won't be included in the output record?

Whatever the case, it must be documented in the plugin so that the user knows what to expect.